src/cobra/core/gene.py

```diff
@@ -559,8 +565,156 @@ def _repr_html__(self):
+        GPRGene_dict: dict = None,
+    ) -> Union[spl.Or, spl.And, Symbol]:
+        """Parse gpr into SYMPY using ast similar to _ast2str().
+
+        Parameters
+        ----------
+        expr : AST or GPR or list or Name or BoolOp
+            compiled GPR
+        GPRGene_dict: dict
+            dictionary from gene id to GPRGeneSymbol
+        Returns
+        -------
+        Symbol or BooleanFunction
+            SYMPY expression (Symbol or And or Or). Symbol("") if the GPR is empty
+        """
+        if GPRGene_dict is None:
+            GPRGene_dict = {gid: Symbol(name=gid) for gid in expr.genes}
+        if isinstance(expr, (Expression, GPR)):
+            return (
+                self._symbolic_gpr(expr.body, GPRGene_dict) if expr.body else Symbol("")
+            )
+        else:
+            if isinstance(expr, Name):
+                return GPRGene_dict.get(expr.id)
+            elif isinstance(expr, BoolOp):
+                op = expr.op
+                if isinstance(op, Or):
+                    # noinspection PyTypeChecker
+                    sym_exp = spl.Or(
+                        *[self._symbolic_gpr(i, GPRGene_dict) for i in expr.values]
+                    )
+                elif isinstance(op, And):
+                    # noinspection PyTypeChecker
+                    sym_exp = spl.And(
+                        *[self._symbolic_gpr(i, GPRGene_dict) for i in expr.values]
+                    )
+                else:
+                    raise TypeError("Unsupported operation " + op.__class__.__name)
+                return sym_exp
+            elif not expr:
+                return Symbol("")
+            else:
+                raise TypeError("Unsupported Expression  " + repr(expr))
+
+    @classmethod
+    def from_symbolic(cls, sympy_gpr: Union[spl.BooleanFunction, Symbol]) -> "GPR":
+        """Construct a GPR from a sympy expression.
+
+        Parameters
+        ----------
+        sympy_gpr: sympy
+            a sympy that describes the gene rules, being a Symbol for single genes
+            or a BooleanFunction for AND/OR relationships
+
+        Returns
+        -------
+        GPR:
+            returns a new GPR while setting  self.body as
+            Parsed AST tree that has the gene rules
+            This function also sets self._genes with the gene ids in the AST
+
+        """
+
+        def _sympy_to_ast(
+            sympy_expr: Union[spl.BooleanFunction, Symbol]
+        ) -> Union[BoolOp, Name]:
+            if sympy_expr.func is spl.Or:
+                return BoolOp(
+                    op=Or(), values=[_sympy_to_ast(i) for i in sympy_expr.args]
+                )
+            elif sympy_expr.func is spl.And:
+                return BoolOp(
+                    op=And(), values=[_sympy_to_ast(i) for i in sympy_expr.args]
+                )
+            elif not sympy_expr.args:
+                return Name(id=sympy_expr.name)
```

Comment:
Doesn't this also have to populate the genes? Or maybe that is done somewhere else implicitly.

Reply:
gpr.update_genes() at the end (line 701) repopulates the genes. I guess I could do it explicitly in _sympy_to_ast, but that would complicate the function and I don't see the point. Might be a bit faster, but not much.